src/Juvix/Compiler/Core/Transformation/LambdaLifting.hs

```diff
@@ -51,6 +56,85 @@ lambdaLiftNode aboveBl top =
           let fApp = mkApps' (mkIdent mempty f) (map NVar freevars)
           return (End fApp)
 
+        goLetRec :: LetRec -> Sem r Recur
+        goLetRec letr = do
+          let defs :: [Node]
+              defs = toList (letr ^. letRecValues)
+              ndefs :: Int
+              ndefs = length defs
+              letRecBinders :: [Info]
+              letRecBinders = Info.getInfoBinders ndefs (letr ^. letRecInfo)
+              bl' :: BinderList Info
+              bl' = BL.prepend letRecBinders bl
+          topSyms :: [Symbol] <- forM defs (const freshSymbol)
+          let recItemsFreeVars :: [(Var, Info)]
+              recItemsFreeVars = mapMaybe helper (toList (mconcatMap freeVarsSet defs))
+                where
+                  -- free vars in each let
+                  -- throw away variables bound in the letrec and shift others
+                  helper :: Var -> Maybe (Var, Info)
+                  helper v
+                    | v ^. varIndex < ndefs = Nothing
+                    | otherwise = Just (set varIndex idx' v, BL.lookup idx' bl)
+                    where
+                      idx' = (v ^. varIndex) - ndefs
+
+              subsCalls :: Node -> Node
+              subsCalls =
+                substs
+                  ( reverse
+                      [ mkApps' (mkIdent' sym) (map (NVar . fst) recItemsFreeVars)
+                        | sym <- topSyms
+                      ]
+                  )
+          -- NOTE that we are first substituting the calls and then performing
+          -- lambda lifting. This is a tradeoff. We have slower compilation but
+          -- slightly faster execution time, since it minimizes the number of
+          -- free variables that need to be passed around.
```

Comment:
To be precise: because of this the complexity is O(n m k) instead of O(n k) where:
- n is the size of the program,
- m is the *total* number of letrecs in the whole program,
- k is the maximum nesting of binders.

k can in practice be treated as a non-negligible constant; m not necessarily, but we won't notice this until we have programs with many letrecs